libyul/YulName.h

```diff
@@ -0,0 +1,230 @@
+ *    defined by their number of inputs and outputs
+ *  - GHOST variables are used for specific transformations in the ``ControlFlowGraph`` and can be added
+ *    using ``addGhost``. These always fall under the 'defined' category.
+ */
+class YulNameRepository
+{
+public:
+	using YulName = size_t;
+
+	/// Decorating a yul dialect builtin function with ``YulName``s.
+	struct BuiltinFunction
+	{
+		YulName name;
+		std::vector<YulName> parameters;
+		std::vector<YulName> returns;
+
+		yul::BuiltinFunction const* data;
+	};
+
+	/// A couple of predefined yul names, most of which are dialect builtins.
+	struct PredefinedHandles
+	{
+		YulName empty {}; ///< Special empty name (corresponding to the empty label).
+		YulName verbatim {}; ///< Verbatim builtin parent name. All verbatims are derived from this.
+		YulName boolType {}; ///< Bool type. Depending on dialect, this may be zero/empty.
+		YulName datasize {}; ///< Datasize builtin.
+		YulName dataoffset {}; ///< Dataoffset builtin.
+		YulName selfdestruct {}; ///< Selfdestruct builtin.
+		YulName memoryguard {}; ///< Memoryguard builtin.
+		YulName eq {}; ///< Equality function.
+		YulName add {}; ///< Addition function.
+		YulName sub {}; ///< Subtraction function.
+		YulName tstore {}; ///< Tstore builtin.
+		YulName defaultType {}; ///< Dialect's default type. May be zero/empty.
+		YulName placeholder_zero {}; ///< Special name ``@ 0`` used in the ``UnusedStoreEliminator``.
+		YulName placeholder_one {}; ///< Special name ``@ 1`` used in the ``UnusedStoreEliminator``.
+		YulName placeholder_thirtytwo {}; ///< Special name ``@ 32`` used in the ``UnusedStoreEliminator``.
+	};
+
+	/// Construct via dialect. It is important that the dialect instance lives at least as long as the name repository
+	/// instance.
+	explicit YulNameRepository(Dialect const& _dialect);
+	~YulNameRepository() = default;
+	YulNameRepository(YulNameRepository&&) = default;
+	YulNameRepository& operator=(YulNameRepository&&) = default;
+	YulNameRepository(YulNameRepository const&) = default;
+	YulNameRepository& operator=(YulNameRepository const&) = default;
+
+	/// Defines a new name based on a label. If the label already was defined, it returns the corresponding YulName
+	/// instead of a new one (``defineName("xyz") == defineName("xyz")``.
+	YulName defineName(std::string_view _label);
+
+	/// Defines a new name based on a parent name. When generating labels, the generated label will be based on the
+	/// parent's (``deriveName(id) != deriveName(id)``).
+	YulName deriveName(YulName _id);
+
+	/// Adds a ghost name (``GHOST[n]``, where ``n`` is the number of times ``addGhost`` was called).
+	/// These are used for switch conditions in the ``ControlFlowGraph`` and do not need a label generated for them,
+	/// as they are already 'defined'.
+	YulName addGhost();
+
+	/// The empty name.
+	static constexpr YulName emptyName() { return 0; }
```

Comment:
Yeah it's definitely something that can go wrong. I was concerned for performance so opted for the raw uint - but probably safety is more important and this is not the bottleneck anyways. I'll add such a struct, then we can also assert easily that the yulname argument belongs to the right repository:)

Reply:
Right. `YulString` was already a class though so I suspect this won't really impact performance.

Reply:
I've added another field to it (`yulNameRepositoryInstanceCount` ... not really satisfied with that name :)) which reflects the repository instance the `YulName` was created from and added corresponding asserts. At first I added a pointer to the creating repository, but the repositories are moved around so the pointer would go invalid. Shared ptr seemed overkill.

Reply:
Ok so the origin checking also means I'll have to update all the ast nodes' origins when (eg) optimizing on a copy of the ast with a copy of the name repository. It's probably not too bad performance wise as its just one more pass over the tree but it does mean another change - that could also be added post-hoc. what do you say?

Reply:
Why would it require a pass over the AST?

The way I see it, the repository is a data structure that should update the origins or all the `YulName`s it owns on copy. That would be in its copy constructor (which maybe should be private and used only via a dedicated clone method to make it harder to copy it accidentally).

This should not be *that* much heavier than a normal copy since it's basically filling out two fields per `YulName` rather than one. When you create a new AST, you'd copy the repo along with it and the update would happen automatically. Unless you're moving it instead and discarding the original AST, but that does not require an update anyway.

> At first I added a pointer to the creating repository, but the repositories are moved around so the pointer would go invalid. Shared ptr seemed overkill.

Well, you could use a `weak_ptr`, which would be safe and not that much overhead (since we never need to `lock()` it), but a numeric ID looks sufficient too. I'd just make it private and modifiable only by the repository. Actually, the value should probably be private as well (only readable via a getter).

Reply:
The repository updating all yulnames it owns would mean that it actually owns them and has references to them - which it doesn't in this implementation. It's more loosely coupled in that the repository gives out keys (the YulNames) which can be used to query the labels (or generate them for that matter). So unless I change that behavior, a pass over the ast is required to update all YulName instances it contains st the corresponding assertions don't immediately fail. What I could imagine is having a `cloneAST(ast) -> (nameRepository, ast)` (perhaps better name tbd) method that does the copying and name updating internally. A copy of the ast would be required either way to make it mutable (provided we go with PR #15258).